src/Storage/StreamWrapper.php

```diff
@@ -0,0 +1,612 @@
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Callhack handler for inspecting our current position in the stream
+     *
+     * @return int
+     */
+    public function stream_tell()
+    {
+        return $this->stream->tell();
+    }
+
+    /**
+     * Callback handler for trying to close an opened directory.
+     *
+     * @return bool
+     */
+    public function dir_closedir()
+    {
+        return false;
+    }
+
+    /**
+     * Callback handler for trying to open a directory.
+     *
+     * @param string $path The url directory to open
+     * @param int $options Whether or not to enforce safe_mode
+     * @return bool
+     */
+    public function dir_opendir($path, $options)
+    {
+        $this->openPath($path);
+        $this->dir_rewinddir();
+        return true;
+    }
+
+    /**
+     * Callback handler for reading an entry from a directory handle.
+     *
+     * @return string
+     */
+    public function dir_readdir()
+    {
+        $object = $this->directoryGenerator->current();
+        if ($object) {
+            $this->directoryGenerator->next();
+            return $object->name();
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * Callback handler for rewind the directory handle.
+     *
+     * @return bool
+     */
+    public function dir_rewinddir()
+    {
+        $this->directoryGenerator = $this->bucket->objects([
+            'prefix' => $this->file,
+            'fields' => 'items/name,nextPageToken'
+        ]);
+        return true;
+    }
+
+    /**
+     * Callback handler for trying to create a directory.
+     *
+     * @param string $path The url directory to creaet
+     * @param int $mode The permissions on the directory
+     * @param int $options Bitwise mask of options
+     * @return bool
+     */
+    public function mkdir($path, $mode, $options)
+    {
+        $client = $this->openPath($path);
+        $this->file = $this->makeDirectory($this->file);
+
+        try {
+            $this->bucket->upload('', [
+                'name' => $this->file
+            ]);
+        } catch (ServiceException $e) {
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Callback handler for trying to move a file or directory.
+     *
+     * @param string $from The URL to the current file
+     * @param string $to The URL of the new file location
+     * @return bool
+     */
+    public function rename($from, $to)
+    {
+        $url = parse_url($to);
+        $destinationBucket = $url['host'];
+        $destinationPath = substr($url['path'], 1);
+
+        $this->dir_opendir($from, []);
+        foreach ($this->directoryGenerator as $file) {
+            $name = $file->name();
+            $newPath = str_replace($this->file, $destinationPath, $name);
+
+            $obj = $this->bucket->object($name);
+            $obj->rename($newPath, ['destinationBucket' => $destinationBucket]);
```

Comment:
Yes we should. Since we could be renaming several files at the same time, would you expect that we stop after the first failure, or try to continue on and return false if any have failed? I'm leaning towards aborting after any failure.